lib/internal/modules/esm/worker.js

```diff
@@ -0,0 +1,222 @@
+'use strict';
+
+const {
+  DataViewPrototypeGetBuffer,
+  Int32Array,
+  PromisePrototypeThen,
+  ReflectApply,
+  SafeWeakMap,
+  TypedArrayPrototypeGetBuffer,
+  globalThis: {
+    Atomics: {
+      and: AtomicsAnd,
+      load: AtomicsLoad,
+      notify: AtomicsNotify,
+      or: AtomicsOr,
+      sub: AtomicsSub,
+      wait: AtomicsWait,
+    },
+  },
+} = primordials;
+const assert = require('internal/assert');
+const { clearImmediate, setImmediate } = require('timers');
+const {
+  hasUncaughtExceptionCaptureCallback,
+} = require('internal/process/execution');
+const {
+  isArrayBuffer,
+  isDataView,
+  isTypedArray,
+} = require('util/types');
+
+// Create this WeakMap in js-land because V8 has no C++ API for WeakMap.
+internalBinding('module_wrap').callbackMap = new SafeWeakMap();
+
+const { isMainThread, workerData, receiveMessageOnPort } = require('worker_threads');
+const {
+  IDLENESS_OF_THREADS,
+  MAIN_THREAD_IS_IDLE,
+  NUMBER_OF_INCOMING_ASYNC_RESPONSES,
+  NUMBER_OF_INCOMING_MESSAGES,
+  WORKER_THREAD_IS_IDLE,
+  WORKER_TO_MAIN_THREAD_NOTIFICATION,
+} = require('internal/modules/esm/shared_constants');
+
+if (isMainThread) { return; } // Needed to pass some tests that happen to load this file on the main thread
+
+const lock = new Int32Array(workerData.lock);
+const { syncCommPort } = workerData;
+
+function transferArrayBuffer(hasError, source) {
+  if (hasError || source == null) return;
+  if (isArrayBuffer(source)) return [source];
+  if (isTypedArray(source)) return [TypedArrayPrototypeGetBuffer(source)];
+  if (isDataView(source)) return [DataViewPrototypeGetBuffer(source)];
+}
+
+function releaseLock() {
+  AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION);
+}
+
+function errorHandler(err, origin = 'unhandledRejection') {
+  releaseLock();
+  process.off('uncaughtException', errorHandler);
+  if (hasUncaughtExceptionCaptureCallback()) {
+    process._fatalException(err);
+    return;
+  }
+  internalBinding('errors').triggerUncaughtException(
+    err,
+    origin === 'unhandledRejection',
+  );
+}
+
+/**
+ * ! Run everything possible within this function so errors get reported.
+ */
+(async function customizedModuleWorker() {
+  const { initializeESM, initializeHooks } = require('internal/modules/esm/utils');
+  let hooks;
+  let preloadScripts;
+  let hasInitializationError = false;
+  let initializationError;
```

Comment:
Why also `hasInitializationError`? (`!!initializationError` → has initializationError)

Reply:
What if `initializationError` is falsy?

Reply:
Do user-land exceptions surface here? It looks like the exceptions would come from node-land.

(If user-land, then sure, your argument makes sense)

Reply:
You can convince yourself with trying e.g. `out/Release/node --loader 'data:text/javascript,throw%200'`.